warp_drive/training/lightning_trainer.py

```diff
@@ -0,0 +1,991 @@
+                )
+                print(f"Saving the '{policy}' torch model to the file: '{filepath}'.")
+
+                torch.save(model.state_dict(), filepath)
+
+    def graceful_close(self):
+        # Delete the sample controller to clear
+        # the random seeds defined in the CUDA memory heap
+        del self.cuda_sample_controller
+        print("Trainer exits gracefully")
+
+    def fetch_episode_states(
+        self,
+        list_of_states=None,  # list of states (data array names) to fetch
+        env_id=0,  # environment id to fetch the states from
+    ):
+        """
+        Step through env and fetch the desired states (data arrays on the GPU)
+        for an entire episode. The trained models will be used for evaluation.
+        """
+        assert 0 <= env_id < self.num_envs
+        assert list_of_states is not None
+        assert isinstance(list_of_states, list)
+        assert len(list_of_states) > 0
+
+        logging.info(f"Fetching the episode states: {list_of_states} from the GPU.")
+        # Ensure env is reset before the start of training, and done flags are False
+        self.cuda_envs.reset_all_envs()
+        env = self.cuda_envs.env
+
+        episode_states = {}
+
+        for state in list_of_states:
+            assert self.cuda_envs.cuda_data_manager.is_data_on_device(
+                state
+            ), f"{state} is not a valid array name on the GPU!"
+            # Note: Discard the first dimension, which is the env dimension
+            array_shape = self.cuda_envs.cuda_data_manager.get_shape(state)[1:]
+
+            # Initialize the episode states
+            episode_states[state] = np.nan * np.stack(
+                [np.ones(array_shape) for _ in range(env.episode_length + 1)]
+            )
+
+        for timestep in range(env.episode_length + 1):
+            # Update the episode states
+            for state in list_of_states:
+                episode_states[state][
+                    timestep
+                ] = self.cuda_envs.cuda_data_manager.pull_data_from_device(state)[
+                    env_id
+                ]
+            # Evaluate policies to compute action probabilities
+            probabilities = self._evaluate_policies()
+
+            # Sample actions
+            self._sample_actions(probabilities)
+
+            # Step through all the environments
+            self.cuda_envs.step_all_envs()
+
+            # Fetch the states when episode is complete
+            if env.cuda_data_manager.pull_data_from_device("_done_")[env_id]:
+                for state in list_of_states:
+                    episode_states[state][
+                        timestep
+                    ] = self.cuda_envs.cuda_data_manager.pull_data_from_device(state)[
+                        env_id
+                    ]
+                break
+
+        return episode_states
+
+    # APIs to integrate with Pytorch Lightning
```

Comment:
@tchaton I have integrated the WarpDrive model with some of the relevant Pytorch Lightning APIs. Could you review this file from here? Thank you.